src/policy/immix/immixspace.rs

```diff
@@ -808,6 +802,18 @@ impl<VM: VMBinding> GCWork<VM> for PrepareBlockState<VM> {
             block.set_state(BlockState::Unmarked);
             debug_assert!(!block.get_state().is_reusable());
             debug_assert_ne!(block.get_state(), BlockState::Marked);
+            // Clear forwarding bits if necessary.
+            if is_defrag_source {
+                if let MetadataSpec::OnSide(side) = *VM::VMObjectModel::LOCAL_FORWARDING_BITS_SPEC {
```

Comment:
This moves the bzero code from `reset_object_mark()` which does bzero unconditionally for the entire chunk to here, which is conditional (`is_defrag_source` and not `Unallocated`), and zeroed at block granularity. Is this an optimization?

Reply:
Yes.  This is an optimisation.  It will still be correct if we zero entire chunks unconditionally.

Previously, it was necessary to clear all forwarding bits of all chunks because `post_copy` assumed all forwarding bits for all chunks were cleared.  Since `post_copy` now unconditionally clears the forwarding bits of new copies, we no longer need to clear the forwarding bits of all chunks.  We only need to ensure all objects that can be moved have their forwarding bits (of the old copies) cleared before copying them.

An alternative solution is letting `PrepareBlockState` clear all forwarding bits of all Allocated and Free chunks so that `post_copy` can still elide the clearing of forwarding bits.  That would be too costly because very few blocks are defrag sources.  If the number of copied objects is small, it should be faster to clear the forwarding bits for less blocks in `PrepareBlockState` and clear the forwarding bits for more objects in `post_copy`.

Reply:
I see. The bulk zeroing was also introduced in d341506b70a4465dbbc05e87d27d0b7bd23844bb. So this PR basically reverts both changes in that commit, and does bulk zeroing in another place to fix the original problem. It is probably worth mentioning the commit d341506b70a4465dbbc05e87d27d0b7bd23844bb in the description (which will be the merge commit message).

Reply:
Sure. I'll mention it when merging.